File: cpp/phone-number/phone_number.cpp
#include "phone_number.h"
#include <algorithm>
#include <cctype>
#include <regex>

namespace phone_number {

phone_number::phone_number(const std::string& input) {
    // Remove all non-digit characters
    std::string digits;
    std::copy_if(input.begin(), input.end(), std::back_inserter(digits),
                 [](char c) { return std::isdigit(c); });
    
    // Check if the number has valid length
    if (digits.length() == 11) {
        if (digits[0] != '1') {
            throw std::domain_error("11-digit number must start with 1");
        }
        // Remove the country code
        digits = digits.substr(1);
    } else if (digits.length() != 10) {
        throw std::domain_error("Phone number must be 10 or 11 digits");
    }
    
    // Validate area code and exchange code
    if (digits[0] == '0' || digits[0] == '1') {
        throw std::domain_error("Area code cannot start with 0 or 1");
    }
    
    if (digits[3] == '0' || digits[3] == '1') {
        throw std::domain_error("Exchange code cannot start with 0 or 1");
    }
    
    cleaned_number = digits;
}

std::string phone_number::number() const {
    return cleaned_number;
}

}  // namespace phone_number

File: python/phone-number/phone_number.py
class PhoneNumber:
    def __init__(self, number):
        # Check for letters first
        if any(char.isalpha() for char in number):
            raise ValueError("letters not permitted")
        
        # Check for punctuations (any non-alphanumeric and non-space character)
        if any(char for char in number if not char.isalnum() and not char.isspace() and char not in '()-.+'):
            raise ValueError("punctuations not permitted")
        
        # Clean the input by removing all non-digit characters
        digits_only = ''.join(char for char in number if char.isdigit())
        
        # Check length
        if len(digits_only) < 10:
            raise ValueError("must not be fewer than 10 digits")
        elif len(digits_only) > 11:
            raise ValueError("must not be greater than 11 digits")
        
        # Handle 11-digit numbers
        if len(digits_only) == 11:
            if digits_only[0] != '1':
                raise ValueError("11 digits must start with 1")
            # Remove the country code
            digits_only = digits_only[1:]
        
        # Validate area code
        if digits_only[0] == '0':
            raise ValueError("area code cannot start with zero")
        if digits_only[0] == '1':
            raise ValueError("area code cannot start with one")
        
        # Validate exchange code
        if digits_only[3] == '0':
            raise ValueError("exchange code cannot start with zero")
        if digits_only[3] == '1':
            raise ValueError("exchange code cannot start with one")
        
        # Store the cleaned number
        self._number = digits_only
    
    @property
    def number(self):
        return self._number
    
    @property
    def area_code(self):
        return self._number[:3]
    
    def pretty(self):
        return f"({self.area_code})-{self._number[3:6]}-{self._number[6:]}"
